File: assignement3/two/MatchMaker.java
import java.io.*;
import java.lang.*;
import java.util.*;

class MatchMaker{
	private ArrayList<String> words = new ArrayList<String>();
	private ArrayList<String> numbers = new ArrayList<String>();
	private HashMap<String, String> numberRepresentation = new HashMap<String, String>(); 
	private static final String[] combos = {"e","jnq", "rwx", "dsy", "ft", 
		"am", "civ", "bku", "lop", "ghz"};
	private ArrayList<String> matchingWords = new ArrayList<String>();

	public MatchMaker(){
		readFromFile();
	}

	public void readFromFile(){
		String str = "";
		String str2 = "";
		try{
			BufferedReader br = new BufferedReader(new FileReader("newdict.txt"));
			BufferedReader br2 = new BufferedReader(new FileReader("numbers.txt"));
			while ((str = br.readLine()) != null){
				words.add(str.toLowerCase());
				numberRepresentation.put(str.toLowerCase(), wordsToNumbers(str.toLowerCase()));
			}


			while ((str2 = br2.readLine()) != null){
				try{
					numbers.add(str2);
				}
				catch(NumberFormatException nbe){
					System.out.println("Wrong");
				}
			}
		}
		catch(FileNotFoundException fileNotFoundException){
			System.out.println("File not found");
		}
		catch(IOException e){
			System.out.println(e);
		}



		for(int i = 0; i < numbers.size(); i++){
			String number = numbers.get(i).trim();
			translateNumber(number, "", 0);
			printMatches(number);
		}
	}

	public String wordsToNumbers(String word){
		String s = "";
		for(int i = 0; i < word.length(); i++){
			for(int y = 0; y < combos.length; y++){
				if(combos[y].contains(String.valueOf(word.charAt(i)))){
					s += y;
				}
			}
		}
		return s;
	}


	public void translateNumber(String number, String translation, int index) {
		if (number.length() == 0) {
			matchingWords.add(translation);
		}
		String firstpart = "";
		String restpart = "";
		while (index < number.length()) {
			firstpart = number.substring(0, ++index);
			for (String match : numberRepresentation.keySet()) {
				if (firstpart.equals(numberRepresentation.get(match))) {
					restpart = number.replaceFirst(firstpart, "");
					translateNumber(restpart, translation+match+" ", 0);
				}
			}
		}
	}

	public void printMatches(String number){
		if(matchingWords.size() > 0){
			for(int i = 0; i < matchingWords.size(); i++){
				System.out.print(number+": "+matchingWords.get(i)+"\n");
			}
		}
		matchingWords.clear();
	}

	public static void main(String[] args){
		MatchMaker m = new MatchMaker();
	}
}

File: assignement3/two/MatchMaker.py
#!/usr/bin/python

import fileinput

class MatchMaker:
	combos = ["e","jnq", "rwx", "dsy", "ft", 
		"am", "civ", "bku", "lop", "ghz"]
	numberRepresentation = {}
	matchingWords = []
	
	def loadFromFile(self):
		wordFile = open("newdict.txt")
		numberFile = open("numbers.txt")
		wordLine = wordFile.readline().lower().strip()
		while wordLine:
			self.wordsToNumber(wordLine)
			wordLine = wordFile.readline().lower().strip()
		wordFile.close()
		numberLine = numberFile.readline().strip()
		while numberLine:
			self.translate(numberLine, "", 0)
			self.printWords(numberLine)
			numberLine = numberFile.readline().strip()
		numberFile.close()
	
	def printWords(self, number):
		for match in self.matchingWords:
			print number+": "+match
			self.matchingWords = []
	
	def wordsToNumber(self, word):
		self.numberRepresentation[word] = ""
		for l in word:
			for c in self.combos:
				if l in c:			
					self.numberRepresentation[word] += str(self.combos.index(c))
					
	def translate(self, number, translation, index):
		if(len(number) == 0):
			self.matchingWords.append(translation)
		while index < len(number):
			index = index+1
			firstPart = number[0 : index]
			for k in self.numberRepresentation:
				if(firstPart == self.numberRepresentation[k]):
					restPart = number[len(firstPart):]
					self.translate(restPart, translation+k+" ",0)

#from MatchMaker import MatchMaker
m = MatchMaker()
m.loadFromFile()
